### PR TITLE
Protocol v2.1 and AWS support with c-lib

### PR DESCRIPTION
**Pull request by nik-markovic**
Implementing [Protocol 2.1](https://docs.iotconnect.io/iotconnect/sdk/message-protocol/device-message-2-1/) Features:

* Composing [Telemetry](https://docs.iotconnect.io/iotconnect/sdk/message-protocol/device-message-2-1/d2c-messages/#Device) messages.
* Parsing [C2D OTA](https://docs.iotconnect.io/iotconnect/sdk/message-protocol/device-message-2-1/c2d-messages/#OTA) messages and providing download details.
* Parsing [C2D Command](https://docs.iotconnect.io/iotconnect/sdk/message-protocol/device-message-2-1/c2d-messages/#Device) messages.
* Composing [OTA and command acknowledgements](https://docs.iotconnect.io/iotconnect/sdk/message-protocol/device-message-2-1/d2c-messages) acknowledgements.
* Parsing HTTP [discovery](https://docs.iotconnect.io/iotconnect/sdk/message-protocol/device-message-2-1/discovery-api/)
and [identity](https://docs.iotconnect.io/iotconnect/sdk/message-protocol/device-message-2-1/identity-api/) Composing [Telemetry](https://docs.iotconnect.io/iotconnect/sdk/message-protocol/device-message-2-1/d2c-messages/#Device) messages.

### CODE REVIEW

File: core/include/iotcl_certs.h
Comment (ylin-witekio) [inline core/include/iotcl_certs.h:1]: Pardon my ignorance but is there a reason why we don't have AWS CA2 and CA4? Or just use the the Starfield one?

Are the end points only signed with CA1 and CA3?

File: core/include/iotcl_certs.h [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+"U5PMCCjjmCXPI6T53iHTfIUJrU6adTrCC2qJeHZERxhlbI1Bjjt/msv0tadQ1wUs\n"\
+"N+gDS63pYaACbvXy8MWy7Vu33PqUXHeeE6V/Uq2V8viTO96LXFvKWlJbYK8U90vv\n"\
+"o/ufQJVtMVT8QtPHRh8jrdkPSHCa2XV4cdFyQzR1bldZwgJcJmApzyMZFo6IQ6XU\n"\
+"5MsI+yMRQ+hDKXJioaldXgjUkK642M4UwtBV8ob2xJNDd2ZhwLnoQdeXeGADbkpy\n"\
+"rqXRfboQnoZsG4q5WTP468SQvvG5\n"\
+"-----END CERTIFICATE-----\n"
+
+// IoT Core MQTT server CA Certificate
+// https://www.amazontrust.com/repository/AmazonRootCA3.pem
+// Self Signed Amazon Root CA2
```
Comment (ylin-witekio) [inline core/include/iotcl_certs.h]: Should this comment say CA3 instead of CA2?

File: .gitignore [hunk 2]
```diff
@@ -7,4 +7,4 @@
 cmake-build-*
 
 # Any makefiles and project files that end up in the root afte rproejct is pulled
```
Comment (lagosantol) [inline .gitignore]: Typo on line 9: "afte rproejct"

File: core/include/iotcl_certs.h (hunk 1 above)
Comment (nik-markovic) [reply core/include/iotcl_certs.h]: That is correct. Thanks! Will fix.

**Comment by nik-markovic** [comment]
@lagosantol Please double-check comments in the iotcl_certs.h file. I will also be adding the pem certificates to the tree soon (sowmewhere). I just remembered that it would be useful.

Edit: and also /tools/cert-generation-self-signed/README.md along with the script and default cert settings in it. The README says this:
When presenting Server CA certificates to your MQTT client, use the following:
- For AWS, use certificates form the [Amazon Trust Repository](https://www.amazontrust.com/repository/):
    - Amazon Root CA 3: If using ECC with ECDHE-ECDSA ciphers.
    - Amazon Root CA 1: If using RSA or ECC with ECDHE-RSA ciphers.
    - Starfield Services Root Certificate Authority - G2: If the device cannot process intermediates as root CAs.
- For Azure, use the 
[DigiCert Global Root G2](https://cacerts.digicert.com/DigiCertGlobalRootG2.crt.pem)
 from [this page](https://www.digicert.com/kb/digicert-root-certificates.htm).

File: core/include/iotcl_certs.h
Comment (nik-markovic) [reply core/include/iotcl_certs.h:1]: Depending on ciphers used, different CA will be presented by the endpoint. The 1 and 3 are "typical".
The customer can always use whichever cert they need and we are defining those two most common ones for convenience.

@lagosantol feeel free to chip in.

**Comment by nik-markovic** [comment]
@sw-ms-imranbhadelia @sw-iot-neeravparasar Please advise on how to proceed regarging https://github.com/avnet-iotconnect/iotc-c-lib/blob/protocol-v2.1/core/src/iotcl_c2d.c#L173C8-L173C36 and similar places in code with IOTCL_C2D_LEGACY_ACK_SUPPORT flag. This is in relation to issue https://awspoc.iotconnect.io/support-info/202402120718630 (make sure to click "Load More" 4 times) .

**Comment by nik-markovic** [comment]
@sw-ms-imranbhadelia @sw-iot-neeravparasar Please advise on https://github.com/avnet-iotconnect/iotc-c-lib/blob/protocol-v2.1/modules/device-rest-api/iotcl_dra_identity.c#L105 in relation to tickets https://awspoc.iotconnect.io/support-info/2024036163515369 and https://awspoc.iotconnect.io/support-info/202311217261209.

**Comment by nik-markovic** [comment]
To all:
I am thinking to remove topic length from this callback:
```c
typedef void (*IotclMqttTransportSend)(const char *topic, size_t topic_len, const char *json_str);
```
along with the tpic lengths in IotclMqttConfig struct.

My original thinking was that if an mqtt client takes topic length as an argument, we could "save" on strlen() processing every time a message is sent, but now I see that paho and mosquitto MQTT clients don't take length as an argument for topic, so it just looks like an unnecessary complication to me and a possiblity to tie the callback directly into the client without a need for a proxy function. Also we're not like gonna send 1000000 messages per second so that strlen is really gonna matter. If that's the case the proxy finction in the user code can be used maybe to optimize this.

Please chip in with your thoughts for or against this idea.

Edit: The changes are now committed.

**Comment by nik-markovic** [comment]
Did not get feedback from the back end team, so decided to use legacy ACK behavior as default for now.

**Pull request merged**